File: evals/tasks/oopif_in_osr.ts
import { EvalFunction } from "@/types/evals";

export const oopif_in_osr: EvalFunction = async ({
  debugUrl,
  sessionUrl,
  v3,
  logger,
}) => {
  // this eval is designed to test whether stagehand can successfully
  // fill a form inside a OOPIF (out of process iframe) that is inside an
  // OSR (open mode shadow) root

  try {
    const page = v3.context.pages()[0];
    await page.goto(
      "https://browserbase.github.io/stagehand-eval-sites/sites/oopif-in-open-shadow-dom/",
    );
    await v3.act({ instruction: "fill 'nunya' into the first name field" });

    const extraction = await v3.extract({
      instruction: "extract the entire page text",
    });

    const pageText = extraction.extraction;

    if (pageText.includes("nunya")) {
      return {
        _success: true,
        message: `successfully filled the form`,
        debugUrl,
        sessionUrl,
        logs: logger.getLogs(),
      };
    }
    return {
      _success: false,
      message: `unable to fill the form`,
      debugUrl,
      sessionUrl,
      logs: logger.getLogs(),
    };
  } catch (error) {
    return {
      _success: false,
      message: `error: ${error.message}`,
      debugUrl,
      sessionUrl,
      logs: logger.getLogs(),
    };
  } finally {
    await v3.close();
  }
};

File: packages/sdk-ruby/evals/tasks/act/oopif_in_osr.rb
# frozen_string_literal: true

# Port of packages/evals/tasks/bench/act/oopif_in_osr.ts

Evals.define_task("oopif_in_osr") do |t|
  # this eval is designed to test whether stagehand can successfully
  # fill a form inside a OOPIF (out of process iframe) that is inside an
  # OSR (open mode shadow) root

  t.page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/oopif-in-open-shadow-dom/")
  t.stagehand.act("fill 'nunya' into the first name field")

  # v3 used schemaless extract; v4 requires a schema.
  # Single-word key to stay clear of the snake_case wire-casing bug (#14).
  extraction = t.stagehand.extract(
    "extract the entire page text",
    schema: {
      "type" => "object",
      "properties" => { "extraction" => { "type" => "string" } },
      "required" => ["extraction"],
      "additionalProperties" => false,
    },
  ).data

  page_text = extraction["extraction"]

  if page_text.include?("nunya")
    next { _success: true, message: "successfully filled the form" }
  end

  { _success: false, message: "unable to fill the form" }
end
